UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -960,6 +962,14 @@ ParseDtb (
     }
   }
 
+  // Post processing: TODO: Need to look into it. Such cross dependency on DT nodes
+  // may not be good idea
+  if (PciRbNode) {
+    DEBUG ((DEBUG_INFO, "ParsePciRootBridge, index :%x\n", index));
+    ParsePciRootBridge (Fdt, PciRbNode, PciEnumDone, RootBridgeCount, GmaStr, &index);
+    DEBUG ((DEBUG_INFO, "After ParsePciRootBridge, index :%x\n", index));
+  }
+
   // Post processing: TODO: Need to look into it. Such cross dependency on DT nodes
   // may not be good idea. Instead have this prop part of RB
   mPciRootBridgeInfo->ResourceAssigned = (BOOLEAN)PciEnumDone;

```

Comment:
This makes "pci-rb" required, or NULL pointer exceptions will start happening. Need to check the spec to see if it is, decide if it should be, and if required, implement this in coreboot. Currently, we don't.

Reply:
yes, pci-rb is mandatory in edk2 UPL because the built-in PciBus driver will consume RootBridgeInfo even in the case ResourceAssigned is TRUE (Only skipping resource re-assignment but PCI enumeration still needed to connect devices/drivers)

Reply:
As of https://github.com/tianocore/edk2/tree/a0ac7cf67afd382b4bd13a402bb6100a435b5b7b, it wasn't strictly required, and I was able to boot into an OS. We weren't setting the `pci-enum-done` parameter, didn't have `pci-rb` and from memory, EDK2's PCI drivers just performed full enumeration on its own. I forget what the 'disable enumeration/scanning' PCD defaults to, but if enumeration is on, it might not even call into the rest of the library?

Reply:
I think it was working in your case (without pci-rb) probably because the ScanForRootBridges() when gUniversalPayloadPciRootBridgeInfoGuid absent:
https://github.com/tianocore/edk2/blob/master/UefiPayloadPkg/Library/PciHostBridgeLib/PciHostBridgeLib.c#L218

This implementation is very basic: for single segment only and might have compatibility issue. Also it will increase UPL execution time.

If bootloader can pass pci-rb FDT nodes that will be more accurate and efficient.

Reply:
Ah, yes. I had forgotten about that line when I wrote my comment above. A single segment is sufficient for client platforms at the moment, and we're ready to accept increased execution time since we don't have the resources to complete support for UPL's PCI in coreboot at the moment, so we've put that on hold. Just like EDK2-based Platform Init doesn't have the resources or need to complete proper FIT loading yet.

Can we mark this as resolved now?

Reply:
yes, ok to me.